kai/incident_store_advanced.py

```diff
@@ -0,0 +1,855 @@
+import pathlib
+import requests
+import os
+import random
+from configparser import ConfigParser
+from enum import Enum
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import datetime
+from typing import Any
+from functools import wraps
+from inspect import signature
+import json
+from urllib.parse import unquote, urlparse
+
+from git import Repo
+
+import psycopg2
+from psycopg2.extensions import connection, cursor
+from psycopg2.extras import DictCursor, DictRow
+
+import tiktoken
+
+from InstructorEmbedding import INSTRUCTOR
+import numpy
+import torch
+
+from report import Report
+
+
+class TrimStrategy(Enum):
+  NONE       = 0
+  TRIM_FRONT = 1
+  TRIM_BACK  = 2
+  TRIM_BOTH  = 3
+
+
+def trim_list(toks: list, max_len: int, trim_strategy: TrimStrategy=TrimStrategy.TRIM_BOTH) -> list:
+  if len(toks) > max_len:
+    num_to_trim = len(toks)-max_len
+
+    match trim_strategy:
+      case TrimStrategy.NONE:
+        raise Exception(f"With NONE trim strategy, too many tokens to embed ({len(toks)} > {max_len})!")
+      case TrimStrategy.TRIM_FRONT:
+        toks = toks[num_to_trim:]
+      case TrimStrategy.TRIM_BACK:
+        toks = toks[:-num_to_trim]
+      case TrimStrategy.TRIM_BOTH:
+        n_front = num_to_trim // 2
+        n_back = num_to_trim - n_front
+
+        toks = toks[n_front:-n_back]
+
+  return toks
+
+
+class EmbeddingProvider(ABC):
+  @abstractmethod
+  def get_embedding(self, inp: str) -> list | None:
+    pass
+
+  @abstractmethod
+  def get_max_tokens(self) -> int:
+    pass
+
+  @abstractmethod
+  def get_dimension(self) -> int:
+    pass
+
+
+class EmbeddingNone(EmbeddingProvider):
+  def __init__(self):
+    self.max_tokens = 128
+    self.dimension  = 1
+
+
+  def get_embedding(self, inp: str) -> list | None:
+    return [0]
+  
+
+  def get_max_tokens(self) -> int:
+    return self.max_tokens
+  
+
+  def get_dimension(self) -> int:
+    return self.dimension
+  
+
+class EmbeddingOpenAI(EmbeddingProvider):
```

Comment:
if we choose to use different model, would this function require any changes? If so, should we make it generic if possible.

Reply:
We can make a new EmbeddingProvider if we want to swap out embeddings, it'd (hopefully :crossed_fingers:) be not too difficult.